x/stakeibc/keeper/transfer.go

```diff
@@ -0,0 +1,214 @@
+	if hostZone.CommunityPoolHoldingAddress == "" || hostZone.CommunityPoolDepositIcaAddress == "" {
+		return errors.New("Invalid holding address or deposit address, cannot build valid ICA commands")
+	}
+
+	// The memo may contain autopilot commands to atomically liquid stake/redeem tokens when transfer succeeds
+	//  both transfer+liquid stake will succeed and tokens will end in the stride side holding address, 
+	//  or neither will and the original base tokens will return to the foreign deposit ICA address
+	memoCommands := ""
+	autopilotMetadata := autopilottypes.RawPacketMetadata{
+		Autopilot: &struct{
+			Receiver string                  				`json:"receiver"`
+			Stakeibc *autopilottypes.StakeibcPacketMetadata `json:"stakeibc,omitempty"`
+			Claim    *autopilottypes.ClaimPacketMetadata    `json:"claim,omitempty"`
+		}{
+			Receiver: hostZone.CommunityPoolHoldingAddress,
+			Stakeibc: &autopilottypes.StakeibcPacketMetadata{
+				Action: "LiquidStake",
+				StrideAddress: hostZone.CommunityPoolHoldingAddress,
+			},
+		},
+	}
+
+	if autoPilotAction == LiquidStake {
+		autopilotMetadata.Autopilot.Stakeibc.Action = "LiquidStake"		
+		autopilotCmdBz, err := json.Marshal(autopilotMetadata)
+		if err != nil {
+			return errorsmod.Wrapf(err, "Couldn't build autopilot json for %v", autopilotMetadata)
+		}
+
+		memoCommands = string(autopilotCmdBz)
+		k.Logger(ctx).Info(fmt.Sprintf("[TransferCommunityPoolTokens] Transferring %v %s and then liquid staking with memo %s", 
+			token.Amount, token.Denom, memoCommands))
+	} else if autoPilotAction == RedeemStake {
+		autopilotMetadata.Autopilot.Stakeibc.Action = "RedeemStake"		
+		autopilotCmdBz, err := json.Marshal(autopilotMetadata)
+		if err != nil {
+			return errorsmod.Wrapf(err, "Couldn't build autopilot json for %v", autopilotMetadata)
+		}
+
+		memoCommands = string(autopilotCmdBz)
+		k.Logger(ctx).Info(fmt.Sprintf("[TransferCommunityPoolTokens] Transferring %v %s and then redeeming stake with memo %s", 
+			token.Amount, token.Denom, memoCommands))
+	} else {
+		k.Logger(ctx).Info(fmt.Sprintf("[TransferCommunityPoolTokens] Transferring %v %s with no additional action", 
+			token.Amount, token.Denom))
+	}
+
+	// get community pool chain's transfer channel for sending tokens from hostZone to Stride
+	transferChannel, found := k.IBCKeeper.ChannelKeeper.GetChannel(ctx, ibctypes.PortID, hostZone.TransferChannelId)
+	if !found {
+		return errorsmod.Wrapf(channeltypes.ErrChannelNotFound, "transfer channel %s not found", hostZone.TransferChannelId)
+	}
+	incomingTransferChannelId := transferChannel.Counterparty.ChannelId
+
+	// one timeout for ICA command ibc messages to arrive, other timeout for subsequent transfer message itself 
+	ibcTransferTimeoutNanos := k.GetParam(ctx, types.KeyIBCTransferTimeoutNanos)
+	icaTimeoutTimestamp := uint64(ctx.BlockTime().UnixNano()) + ibcTransferTimeoutNanos
+	transferTimeoutTimestamp := icaTimeoutTimestamp + ibcTransferTimeoutNanos
```

Comment:
Ok, that certainly simplifies things.  The weird case is if the ICQ is triggered at the start of the epoch and takes almost the whole epoch before the callback kicks off the transfer so instead of a fixed time window it hits the epoch end very shortly after it is kicked off -- but given stride epoch length this seems unlikely and even if it fails the tokens will be processed in the next epoch trigger anyway

Reply:
Yeah that's definitely the relevant edge case, but I think it's pretty unlikely and even if it does occur, we can recover from it next epoch anyways like you said